ext/opcache/jit/zend_jit_x86.dasc

```diff
@@ -0,0 +1,9955 @@
+||				} else {
+|					GET_ZVAL_PTR Ra(tmp_reg2), src_addr
+|					GET_ZVAL_W2 Ra(tmp_reg1), src_addr
+|					SET_ZVAL_PTR dst_addr, Ra(tmp_reg2)
+|					SET_ZVAL_PTR res_addr, Ra(tmp_reg2)
+|					SET_ZVAL_W2 dst_addr, Ra(tmp_reg1)
+|					SET_ZVAL_W2 res_addr, Ra(tmp_reg1)
+||				}
+|			.endif
+||		}
+||	}
+||	if ((src_info & (MAY_BE_NULL|MAY_BE_FALSE|MAY_BE_TRUE|MAY_BE_LONG|MAY_BE_DOUBLE)) &&
+||	    has_concrete_type(src_info & MAY_BE_ANY)) {
+||		zend_uchar type = concrete_type(src_info);
+||		if (Z_MODE(dst_addr) == IS_MEM_ZVAL) {
+||			if (((dst_info & MAY_BE_ANY) != (1<<type)) || (dst_info & (MAY_BE_STRING|MAY_BE_ARRAY))) {
+|				SET_ZVAL_TYPE_INFO dst_addr, type
+||			}
+||		}
+||		if (Z_MODE(res_addr) == IS_MEM_ZVAL) {
+|			SET_ZVAL_TYPE_INFO res_addr, type
+||		}
+||	} else {
+|		GET_ZVAL_TYPE_INFO Rd(tmp_reg1), src_addr
+|		SET_ZVAL_TYPE_INFO dst_addr, Rd(tmp_reg1)
+|		SET_ZVAL_TYPE_INFO res_addr, Rd(tmp_reg1)
+||	}
+|.endmacro
+
+|.macro IF_TYPE, type, val, label
+|	cmp type, val
+|	je label
+|.endmacro
+
+|.macro IF_NOT_TYPE, type, val, label
+|	cmp type, val
+|	jne label
+|.endmacro
+
+|.macro IF_Z_TYPE, zv, val, label
+|	IF_TYPE byte [zv + offsetof(zval, u1.v.type)], val, label
+|.endmacro
+
+|.macro IF_NOT_Z_TYPE, zv, val, label
+|	IF_NOT_TYPE byte [zv + offsetof(zval, u1.v.type)], val, label
+|.endmacro
+
+|.macro CMP_ZVAL_TYPE, addr, val
+||	ZEND_ASSERT(Z_MODE(addr) == IS_MEM_ZVAL);
+|	cmp byte [Ra(Z_REG(addr))+Z_OFFSET(addr)+offsetof(zval, u1.v.type)], val
+|.endmacro
+
+|.macro IF_ZVAL_TYPE, addr, val, label
+||	ZEND_ASSERT(Z_MODE(addr) == IS_MEM_ZVAL);
+|	IF_TYPE byte [Ra(Z_REG(addr))+Z_OFFSET(addr)+offsetof(zval, u1.v.type)], val, label
+|.endmacro
+
+|.macro IF_NOT_ZVAL_TYPE, addr, val, label
+||	ZEND_ASSERT(Z_MODE(addr) == IS_MEM_ZVAL);
+|	IF_NOT_TYPE byte [Ra(Z_REG(addr))+Z_OFFSET(addr)+offsetof(zval, u1.v.type)], val, label
+|.endmacro
+
+|.macro IF_FLAGS, type_flags, mask, label
+|	test type_flags, mask
+|	jnz label
+|.endmacro
+
+|.macro IF_NOT_FLAGS, type_flags, mask, label
+|	test type_flags, mask
+|	jz label
+|.endmacro
+
+|.macro IF_REFCOUNTED, type_flags, label
+|	IF_FLAGS type_flags, IS_TYPE_REFCOUNTED, label
+|.endmacro
+
+|.macro IF_NOT_REFCOUNTED, type_flags, label
+|	IF_NOT_FLAGS type_flags, IS_TYPE_REFCOUNTED, label
+|.endmacro
+
+|.macro IF_ZVAL_FLAGS, addr, mask, label
+||	ZEND_ASSERT(Z_MODE(addr) == IS_MEM_ZVAL);
+|	IF_FLAGS byte [Ra(Z_REG(addr))+Z_OFFSET(addr)+offsetof(zval, u1.v.type_flags)], mask, label
+|.endmacro
+
+|.macro IF_NOT_ZVAL_FLAGS, addr, mask, label
+||	ZEND_ASSERT(Z_MODE(addr) == IS_MEM_ZVAL);
+|	IF_NOT_FLAGS byte [Ra(Z_REG(addr))+Z_OFFSET(addr)+offsetof(zval, u1.v.type_flags)], mask, label
+|.endmacro
+
+|.macro IF_ZVAL_REFCOUNTED, addr, label
+|	IF_ZVAL_FLAGS addr, IS_TYPE_REFCOUNTED, label
+|.endmacro
+
+|.macro IF_NOT_ZVAL_REFCOUNTED, addr, label
+|	IF_NOT_ZVAL_FLAGS addr, IS_TYPE_REFCOUNTED, label
+|.endmacro
+
+|.macro GC_ADDREF, zv
+|	add	dword [zv], 1
```

Comment:
This is on purpose. inc/dec are bad instructions, because they change CPU flags register **partially**, that may cause some penalties. Intel optimization manual recommends to use add/sub.